### PR TITLE
Fix von mises

### PR DESCRIPTION
**Pull request by RiccardoRossi**
**Description**
the formula for Von Mises in 2D was wrong.
this is done in the base_solid_element however @loumalouomega @AlejandroCornejo this error may be present elsewhere so please take a look 

Please mark the PR with appropriate tags: 
- FastPR

**Changelog**
correct  the von mises formula

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp [hunk 1]
```diff
@@ -899,12 +899,15 @@ void BaseSolidElement::CalculateOnIntegrationPoints(
                 double sigma_equivalent = 0.0;
 
                 if (dimension == 2) {
-                    sigma_equivalent = std::pow((stress_tensor(0,0) - stress_tensor(1,1)), 2.0) +
-                                                3*(stress_tensor(0,1) * stress_tensor(1,0));
+
+                    sigma_equivalent = std::pow(stress_tensor(0,0),2)
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: @loumalouomega @AlejandroCornejo, not for this PR, however please note that here a stress_tensor is allocated. this should be best avoided (allocation is by far the most expensive thing we do here) by using the voigt form which we have just above, without converting it into a tensor

**Review by AlejandroCornejo — COMMENTED** [review]
I think that if we are using the VonMises calculation we use the one I implemented inside the `von_mises_yield_surface.h` and not this method. What do you think @loumalouomega ?

**Comment by loumalouomega** [comment]
> I think that if we are using the VonMises calculation we use the one I implemented inside the `von_mises_yield_surface.h` and not this method. What do you think @loumalouomega ?

I was going to say the same. To put that inside CL utilities and call it both from the law and the element.

**Comment by RiccardoRossi** [comment]
ok for moving it to an utility, and taking the occasion to avoid the allocation of the tensor, however i would merge this as a quick fix and have the porting done in another PR (can i ask you guys to take care of that?)

**Comment by loumalouomega** [comment]
> ok for moving it to an utility, and taking the occasion to avoid the allocation of the tensor, however i would merge this as a quick fix and have the porting done in another PR (can i ask you guys to take care of that?)

Don't do it, I am doing it right now

**Comment by loumalouomega** [comment]
One minute

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp [hunk 2]
```diff
@@ -894,29 +895,19 @@ void BaseSolidElement::CalculateOnIntegrationPoints(
...
-
-                double sigma_equivalent = 0.0;
-
                 if (dimension == 2) {
-
-                    sigma_equivalent = std::pow(stress_tensor(0,0),2) 
-                                     + std::pow(stress_tensor(1,1),2) 
-                                     - stress_tensor(0,0)*stress_tensor(1,1) 
-                                     + 3*(stress_tensor(0,1) * stress_tensor(1,0));
+                    array_1d<double, 3> stress_vector;
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: @loumalouomega why don't you template the function in the vector type? this way you can avoid this copy.

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp [hunk 3]
```diff
@@ -894,29 +895,19 @@ void BaseSolidElement::CalculateOnIntegrationPoints(
...
+                    }
+                    ConstitutiveLawUtilities<3>::CalculateVonMisesEquivalentStress(stress_vector, this_constitutive_variables.StrainVector, rOutput[point_number]);
                 } else {
-                    sigma_equivalent = 0.5*(std::pow((stress_tensor(0,0) - stress_tensor(1,1)), 2) +
-                                            std::pow((stress_tensor(1,1) - stress_tensor(2,2)), 2) +
-                                            std::pow((stress_tensor(2,2) - stress_tensor(0,0)), 2) +
-                                                    6*(stress_tensor(0,1) * stress_tensor(1,0) +
-                                                        stress_tensor(1,2) * stress_tensor(2,1) +
-                                                        stress_tensor(2,0) * stress_tensor(0,2)));
+                    array_1d<double, 6> stress_vector;
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: same

File: applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp [hunk 4]
```diff
@@ -1117,6 +1117,25 @@ void ConstitutiveLawUtilities<3>::CalculateRotationOperatorVoigt(
...
+void ConstitutiveLawUtilities<TVoigtSize>::CalculateVonMisesEquivalentStress(
+    const array_1d<double, VoigtSize>& rStressVector,
+    const Vector& rStrainVector,
+    double& rEquivalentStress
+    )
+{
+    double I1, J2;
+    array_1d<double, VoigtSize> deviator = ZeroVector(VoigtSize);
+
+    ConstitutiveLawUtilities<VoigtSize>::CalculateI1Invariant(rStressVector, I1);
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp]: those two functions should also be templated in the vector_type

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp (hunk 2 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: Is for the original CL which uses array

File: applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp (hunk 4 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp]: changing

**Comment by loumalouomega** [comment]
@RiccardoRossi what you propose means to refactor all the CL implemented with the static composition (and it was done for performance reasons), I think it is not the moment for that. If you see in the CL utilities the templates are explicitly specialized

**Comment by loumalouomega** [comment]
> @RiccardoRossi what you propose means to refactor all the CL implemented with the static composition (and it was done for performance reasons), I think it is not the moment for that. If you see in the CL utilities the templates are explicitly specialized

BTW, the implementation you corrected created an auxiliar matrix, copying even more values. This is even less copy-prone than the previous format

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp [hunk 5]
```diff
@@ -887,33 +888,28 @@ void BaseSolidElement::CalculateOnIntegrationPoints(
 
             Values.SetStrainVector(this_constitutive_variables.StrainVector);
 
+            const std::function<void(const Vector&, const Vector&, double&)> retrieve_vm_stress_2d =
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: why do you need a function here?

a very simple solution is to template the function CaclulateVonMisesEquivalentStress in the InputVectoryTypes and then just call it. this way you can avoid any copying.

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp [hunk 6]
```diff
@@ -894,26 +895,12 @@ void BaseSolidElement::CalculateOnIntegrationPoints(
...
-                const Matrix stress_tensor = MathUtils<double>::StressVectorToTensor( this_constitutive_variables.StressVector );
-
-                double sigma_equivalent = 0.0;
-
-                if (dimension == 2) {
-                    sigma_equivalent = std::pow((stress_tensor(0,0) - stress_tensor(1,1)), 2.0) +
-                                                3*(stress_tensor(0,1) * stress_tensor(1,0));
+                // Compute VM stress
+                if (dimension == 2 ) {
+                    ConstitutiveLawUtilities<3>::CalculateVonMisesEquivalentStress(this_constitutive_variables.StressVector, this_constitutive_variables.StrainVector, rOutput[point_number]);
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: why do you need the StrainVector in here???

File: applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.h [hunk 7]
```diff
@@ -95,6 +95,27 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ConstitutiveLawUtilities
     /// The zero tolerance
     static constexpr double tolerance = std::numeric_limits<double>::epsilon();
 
+    #define CalculateJ2Invariant3D                                 \
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.h]: this is REALLY hacky

**Review by RiccardoRossi — COMMENTED** [review]
this looks good to me! i cannot be the one approving, but i think this solution is not clean

**Comment by RiccardoRossi** [comment]
this solution is NOW clean (not->NOW)

**Review by loumalouomega — APPROVED** [review]
I will approve myself (evil laugh)

**Comment by loumalouomega** [comment]
> this solution is NOW clean (not->NOW)

You can edit your comments :)

**Pull request merged**